packages/nameguard-sdk/src/index.ts

```diff
@@ -590,37 +621,50 @@ export class NameGuard {
    * If this resolution fails then NameGuard will return an error.
    *
    * @param {string} namehash A namehash should be a decimal or a hex (prefixed with 0x) string.
-   * @param {InspectNamehashOptions} options The options for the inspection.
    * @returns {Promise<NameGuardReport>} A promise that resolves with the `NameGuardReport` of the name. Check the `inspected` field of the result to determine if the result is an `InspectedNameGuardReport` or an `UninspectedNameGuardReport` for performance reasons in the case that the provided `name` was exceptionally long.
    * @throws {NameGuardError} If the inspection fails due to network issues, server errors, or if the namehash cannot be resolved to a name.
    * @throws {Error} If the provided `namehash` is not in a valid Keccak256 hash format.
    */
 
-  public async inspectNamehash(
-    namehash: string,
-    options?: InspectNamehashOptions,
-  ): Promise<NameGuardReport> {
+  public async inspectNamehash(namehash: string): Promise<NameGuardReport> {
     if (!isKeccak256Hash(namehash)) {
-      throw new Error("Invalid Keccak256 hash format for namehash.");
+      throw new NameGuardError(
+        NameGuardErrorType.Input,
+        undefined,
+        "Invalid Keccak256 hash format for namehash.",
+      );
     }
 
     const network = this.network;
 
-    const url = new URL(
-      `inspect-namehash/${network}/${namehash}`,
-      this.endpoint,
-    );
+    try {
+      const url = new URL(
+        `inspect-namehash/${network}/${namehash}`,
+        this.endpoint,
+      );
+
+      const response = await fetch(url);
 
-    const response = await fetch(url);
+      if (!response.ok) {
+        throw new NameGuardError(
+          NameGuardErrorType.Server,
+          response.status,
+          `Failed to inspect namehash ${namehash} using the network ${network}.`,
+        );
+      }
+
+      return await response.json();
+    } catch (error) {
+      if (error instanceof NameGuardError) {
+        throw error;
+      }
 
-    if (!response.ok) {
       throw new NameGuardError(
-        response.status,
-        `Failed to inspect namehash ${namehash} using the network ${network}.`,
+        NameGuardErrorType.Unknown,
```

Comment:
We should add handling to each of these network calls that detects specific error scenarios such as server / network / timeout / abort.

We should also look at the API docs for each NameGuard API and use the related HTTP status codes to help return more specific error types.

Reply:
What if we did this (keep in mind not everything here regarding response codes is in the openapi spec that I looked through) but we could remove any from the example below that we don't want to add (Rate limiting for example):

``ts
async rawRequest(
  path: string,
  method: string = "GET",
  body: object = {},
  headers: object = {}
): Promise<any> {
  const url = new URL(path, this.endpoint);

  const options: RequestInit = {
    method,
    headers: {
      "Content-Type": "application/json",
      ...headers,
    },
    signal: this.abortController.signal,
  };

  if (method !== "GET") {
    options.body = JSON.stringify(body);
  }

  try {
    const response = await fetch(url, options);

    if (!response.ok) {
      // Handle specific HTTP status codes
      switch (response.status) {
        case 400:
          throw new NameGuardError(
            NameGuardErrorType.Input,
            response.status,
            "Bad request: The server could not understand the request"
          );
        case 401:
          throw new NameGuardError(
            NameGuardErrorType.Authentication,
            response.status,
            "Unauthorized: Authentication is required and has failed or has not been provided"
          );
        case 403:
          throw new NameGuardError(
            NameGuardErrorType.Authorization,
            response.status,
            "Forbidden: The server understood the request but refuses to authorize it"
          );
        case 404:
          throw new NameGuardError(
            NameGuardErrorType.NotFound,
            response.status,
            "Not Found: The requested resource could not be found"
          );
        case 429:
          throw new NameGuardError(
            NameGuardErrorType.RateLimit,
            response.status,
            "Too Many Requests: You have sent too many requests in a given amount of time"
          );
        case 500:
          throw new NameGuardError(
            NameGuardErrorType.Server,
            response.status,
            "Internal Server Error: The server has encountered a situation it doesn't know how to handle"
          );
        default:
          if (response.status >= 500) {
            throw new NameGuardError(
              NameGuardErrorType.Server,
              response.status,
              `Server error: ${response.statusText}`
            );
          } else {
            throw new NameGuardError(
              NameGuardErrorType.Unknown,
              response.status,
              `Unexpected error: ${response.statusText}`
            );
          }
      }
    }

    return await response.json();
  } catch (error) {
    if (error instanceof NameGuardError) {
      throw error;
    } else if (error instanceof TypeError && error.message === "Failed to fetch") {
      throw new NameGuardError(
        NameGuardErrorType.Network,
        undefined,
        "Network error: Unable to reach the server"
      );
    } else if (error instanceof DOMException && error.name === "AbortError") {
      throw new NameGuardError(
        NameGuardErrorType.Abort,
        undefined,
        "Request was aborted"
      );
    } else {
      throw new NameGuardError(
        NameGuardErrorType.Unknown,
        undefined,
        `An unexpected error occurred: ${error instanceof Error ? error.message : String(error)}`
      );
    }
  }
}
```

Reply:
@notrab This looks like a good direction, thank you :+1:

A few suggestions for feedback:

1. What about the timeout case? Can you please simulate a timeout and verify our logic can detect that case correctly?
2. I'm concerned about the `DOMException` referenced in the code above. Can you please suggest a path forward for us where we can be sure all the code here works equally well both in frontend contexts where it runs in the browser and in backend contexts where it runs in node?
3. Suggest we reduce the number of `NameGuardErrorType` values to only those we have real use cases for currently. For example, I understand we never have errors for `NameGuardErrorType.Authentication`, `NameGuardErrorType.Authorization`, or `NameGuardErrorType.RateLimit`. We can introduce new error codes in the future if / when we build related features for that. What do you think? A key motivation for this suggestion is that each error code case introduces the possibility of needing specialized UX handling in the frontend. Maybe I'm being too cautious about that. I'm coming at this from the perspective of trying to avoid introducing too many UI states. But perhaps that's not an issue and we can create just a single generic UI wherever error state reporting into the UI is needed.